File: java/src/main/java/com/swim/smime/AmqpConsumer.java
package com.swim.smime;

import org.apache.qpid.protonj2.client.*;

import java.nio.file.Path;
import java.security.PrivateKey;
import java.security.cert.X509Certificate;

/**
 * AMQP Consumer: subscribes to an AMQP 1.0 queue using native AMQP 1.0 client (ProtonJ2),
 * receives S/MIME messages, verifies signatures and/or decrypts them.
 * Demonstrates SWIM-TIYP-0112 over AMQP with Message Security binding.
 */
public class AmqpConsumer {

    public static void main(String[] args) throws Exception {
        String certsDir = args.length > 0 ? args[0] : "../certs";
        String brokerHost = args.length > 1 ? args[1] : "localhost";
        int brokerPort = args.length > 2 ? Integer.parseInt(args[2]) : 5672;
        String queueName = args.length > 3 ? args[3] : "swim.flight.data";

        System.out.println("=== SWIM S/MIME AMQP Consumer (Java) ===");
        System.out.println("Broker: " + brokerHost + ":" + brokerPort);
        System.out.println("Queue: " + queueName);
        System.out.println("Waiting for messages... (press Ctrl+C to exit)");

        PrivateKey consumerKey = SmimeHelper.loadPrivateKey(
                Path.of(certsDir, "consumer.pfx").toString(), "changeit");
        X509Certificate producerCert = SmimeHelper.loadCertificate(
                Path.of(certsDir, "producer.crt").toString());

        Client client = Client.create();
        ConnectionOptions connOpts = new ConnectionOptions();
        connOpts.user("admin");
        connOpts.password("admin");

        try (Connection connection = client.connect(brokerHost, brokerPort, connOpts);
             Receiver receiver = connection.openReceiver(queueName)) {

            while (true) {
                Delivery delivery = receiver.receive();
                Message<?> msg = delivery.message();

                System.out.println("\n=== Received AMQP Message ===");

                // SWIM-TIYP-0112: read content-type from AMQP core property
                String contentType = msg.contentType();
                // SWIM-TIYP-0112: read MIME-Version from application property (with hyphen)
                Object mimeVersion = msg.property("MIME-Version");

                System.out.println("Content-Type: " + contentType);
                System.out.println("MIME-Version: " + mimeVersion);

                byte[] body;
                Object bodyObj = msg.body();
                if (bodyObj instanceof byte[] bytes) {
                    body = bytes;
                } else {
                    body = bodyObj.toString().getBytes("UTF-8");
                }
                System.out.println("Body size: " + body.length + " bytes");

                try {
                    byte[] payload;

                    if (contentType != null && contentType.contains("smime-type=signed-data")) {
                        payload = SmimeHelper.verifySignature(body, contentType, producerCert);
                        System.out.println("Signature VERIFIED successfully");

                    } else if (contentType != null && contentType.contains("smime-type=enveloped-data")) {
                        payload = SmimeHelper.decryptAndVerify(body, contentType,
                                consumerKey, producerCert);
                        System.out.println("Message DECRYPTED and signature VERIFIED");

                    } else {
                        System.err.println("Unsupported Content-Type: " + contentType);
                        continue;
                    }

                    System.out.println("Extracted payload:\n" + new String(payload, "UTF-8").trim());
                    System.out.flush();

                } catch (Exception e) {
                    System.err.println("Error processing message: " + e.getMessage());
                    e.printStackTrace();
                }
            }
        }
    }
}

File: python/amqp_consumer.py
"""
AMQP Consumer (Python): subscribes to an AMQP 1.0 queue, receives S/MIME messages,
verifies signatures and/or decrypts them.
Demonstrates SWIM-TIYP-0112 over AMQP with Message Security binding.
"""

import os
import sys
import argparse

from smime_helper import (
    load_certificate, load_private_key,
    verify_signature, decrypt_and_verify
)


def main():
    parser = argparse.ArgumentParser(description="SWIM S/MIME AMQP Consumer (Python)")
    parser.add_argument("--certs-dir", default=os.path.join(os.path.dirname(__file__), "..", "certs"))
    parser.add_argument("--broker", default="localhost:5672")
    parser.add_argument("--queue", default="swim.flight.data")
    args = parser.parse_args()

    print("=== SWIM S/MIME AMQP Consumer (Python) ===")
    print(f"Broker: {args.broker}")
    print(f"Queue: {args.queue}")

    producer_cert = load_certificate(os.path.join(args.certs_dir, "producer.crt"))
    consumer_key = load_private_key(os.path.join(args.certs_dir, "consumer.pfx"), "changeit")
    consumer_cert = load_certificate(os.path.join(args.certs_dir, "consumer.pfx"), "changeit")

    try:
        from proton.handlers import MessagingHandler
        from proton.reactor import Container

        class SmimeReceiver(MessagingHandler):
            def __init__(self, broker, queue):
                super().__init__()
                self.broker = broker
                self.queue = queue

            def on_start(self, event):
                conn = event.container.connect(self.broker,
                                                user="admin", password="admin",
                                                allowed_mechs="PLAIN")
                event.container.create_receiver(conn, self.queue)
                print("Waiting for messages... (press Ctrl+C to exit)")

            def on_message(self, event):
                msg = event.message
                print(f"\n=== Received AMQP Message ===")

                # SWIM-TIYP-0112: read content-type from AMQP core property
                content_type = msg.content_type or ""
                # SWIM-TIYP-0112: read MIME-Version from application property (with hyphen)
                mime_version = (msg.properties or {}).get("MIME-Version", "")
                print(f"Content-Type: {content_type}")
                print(f"MIME-Version: {mime_version}")

                body = msg.body
                if isinstance(body, str):
                    body = body.encode("utf-8")
                elif isinstance(body, bytes):
                    pass
                else:
                    body = bytes(body) if body else b""
                print(f"Body size: {len(body)} bytes")

                try:
                    if "smime-type=signed-data" in content_type:
                        payload = verify_signature(body, content_type, producer_cert)
                        print("Signature VERIFIED successfully")
                    elif "smime-type=enveloped-data" in content_type:
                        payload = decrypt_and_verify(body, content_type,
                                                      consumer_key, consumer_cert,
                                                      producer_cert)
                        print("Message DECRYPTED and signature VERIFIED")
                    else:
                        print(f"Unsupported Content-Type: {content_type}")
                        return

                    print(f"Extracted payload:\n{payload.decode('utf-8').strip()}")

                except Exception as e:
                    print(f"Error processing message: {e}", file=sys.stderr)
                    import traceback
                    traceback.print_exc()

        Container(SmimeReceiver(args.broker, args.queue)).run()

    except ImportError:
        print("ERROR: python-qpid-proton is not available on Windows.")
        print("Using fallback AMQP consumer with 'python-qpid-proton' alternative...")
        _run_fallback_consumer(args, producer_cert, consumer_key, consumer_cert)


def _run_fallback_consumer(args, producer_cert, consumer_key, consumer_cert):
    """Fallback AMQP consumer using uamqp or manual approach."""
    print("NOTE: On Windows, qpid-proton may not be available.")
    print("Install it via: pip install python-qpid-proton")
    print("Or use WSL / Linux for the AMQP Python components.")
    sys.exit(1)


if __name__ == "__main__":
    main()
